rust/processor/src/parquet_manager.rs

```diff
@@ -0,0 +1,554 @@
+    Transaction(crate::models::default_models::parquet_transactions::Transaction),  
+    MoveResource(crate::models::default_models::parquet_move_resources::MoveResource),
+    WriteSetChange(crate::models::default_models::parquet_write_set_changes::WriteSetChange),
+}
+
+impl ParquetManager {
+    pub fn new(bucket_name: String) -> Self {
+        Self {
+            writer: None,
+            buffer: Vec::new(),
+            buffer_size_bytes: 0,
+            transaction_version_to_struct_count: AHashMap::new(),
+            bucket_name,
+        }
+    }
+
+    pub fn update_buffer_size(&mut self, parquet_struct: &ParquetStruct) {
+        match parquet_struct {
+            ParquetStruct::MoveResource(move_resource) => {
+                self.buffer_size_bytes += move_resource.size_of();
+            },
+            ParquetStruct::Transaction(transaction) => {
+                self.buffer_size_bytes += transaction.size_of();
+            },
+            ParquetStruct::WriteSetChange(write_set_change) => {
+                self.buffer_size_bytes += write_set_change.size_of();
+            },
+        };
+        debug!("buffer size after adding struct to buffer : {} bytes.", self.buffer_size_bytes);
+    }
+
+    pub async fn write_and_upload(
+        &mut self, 
+        changes: ParquetData, 
+        gcs_client: &GCSClient, 
+        gap_detector_sender: kanal::AsyncSender<ParquetProcessingResult>) -> Result<ParquetProcessingResult> {
+    
+        let parquet_structs = changes.data.first();
+        // update the transaction counter? 
+        // from map in the data, we have 
+        let _processing_result: Result<ParquetProcessingResult> = match parquet_structs.unwrap().clone() {
+            ParquetStruct::MoveResource(_move_resource) => {
+                return self.handle_move_resource(gcs_client, &gap_detector_sender, changes.clone()).await;
+            },
+            ParquetStruct::Transaction(_transaction) => {
+                return self.handle_transaction(gcs_client, &gap_detector_sender, changes.clone()).await;
+            },
+            // ParquetStruct::WriteSetChange(write_set_change) => {
+            //     return self.handle_move_resource(&gcs_client, &gap_detector_sender, changes.clone()).await;
+            // },
+            _ => {
+                error!("No size found for the parquet struct");
+                panic!("No size found for the parquet struct");
+            }
+        };
+    }
+
+    async fn handle_transaction(
+        &mut self,
+        gcs_client: &GCSClient,
+        gap_detector_sender: &kanal::AsyncSender<ParquetProcessingResult>, 
+        changes: ParquetData) -> Result<ParquetProcessingResult>
+    {
+        let processing_start: std::time::Instant = std::time::Instant::now();
+
+        let props = WriterProperties::builder()
+            .set_compression(parquet::basic::Compression::LZ4) 
+            .build();
+        let props_arc = Arc::new(props.clone());
```

Comment:
you do this even if self.props_writer is some- when is it none? is it only during instantiation?